In [1]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [2]:
# Load the given datasets
categorical_data = pd.read_csv("chronic_kidney_disease_categorical.csv")
numerical_data = pd.read_csv("chronic_kidney_disease_numerical.csv")


# Print the data
print(categorical_data.head())
print(numerical_data.head())

   unique_id   al   su       rbc        pc         pcc          ba  htn   dm  \
0     224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes  yes   
1     992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   no   
2     308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   no   
3     450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   no   
4     881763  0.0  0.0    normal    normal  notpresent  notpresent   no   no   

  cad appet   pe  ane  Target  
0  no  good  yes   no     ckd  
1  no  good   no   no  notckd  
2  no  good  yes   no     ckd  
3  no  good   no  yes     ckd  
4  no  good   no   no  notckd  
   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  7300.0   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  7200.0   
2     421471  37.0  60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  5500.0   
3     76411

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: 
Features are the cols, used to describe the patient's condition. Features are categorical like yes/no or types of something and they can also be numerical like measurements.

Labels are what you are trying to predict. The label is a class, indicating whether a patient has Chronic Kidney Disease or not.

In the dataset, we have both categorical and numerical features.


Categorical features: rbc, htn, appet

Numerical features: age, bp, hemo

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [3]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
print(f"Numerical duplicates: {numerical_data.duplicated().sum()}")

# Drop these duplicate rows
numerical_data.drop_duplicates(inplace=True)

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
print(f"Categorical duplicates: {categorical_data.duplicated().sum()}")
categorical_data.drop_duplicates(inplace=True)

Numerical duplicates: 30
Categorical duplicates: 30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

In [4]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged_data = pd.merge(numerical_data, categorical_data, on="unique_id", how="outer")
missing_ids = merged_data[merged_data.isnull().any(axis=1)]
print(missing_ids)
combined_data = pd.merge(numerical_data, categorical_data, on="unique_id")

#Print the combined dataset
print(combined_data.head())


     unique_id   age     bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
1       103570  24.0   80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2       104675  65.0   80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3       105923  90.0   90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4       107297  60.0   70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   
5       108785  53.0  100.0  213.0   23.0   1.0  139.0  4.0   NaN   NaN  ...   
..         ...   ...    ...    ...    ...   ...    ...  ...   ...   ...  ...   
392     991374  34.0   70.0    NaN  219.0  12.2  130.0  3.8   6.0   NaN  ...   
393     992643  35.0   70.0   82.0   36.0   1.1  150.0  3.5  14.5  52.0  ...   
397     996412  70.0   60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0  ...   
398     997629  35.0    NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0  ...   
399     997876  64.0   70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0  ...   

         pc         pcc          ba  ht

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [5]:
# Calculate the percentage of rows that contain atleast one missing value
missing_percentage = combined_data.isnull().mean() * 100
# Print %
print(f"Percentage of rows with missing values: {missing_percentage.mean()}")
# Drop these rows from the dataset
cleaned_data = combined_data.dropna()
# Print the Dataset
print(cleaned_data)

Percentage of rows with missing values: 9.661538461538463
     unique_id   age     bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       938027  43.0   60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
2       421471  37.0   60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0  ...   
7       109053  64.0   60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
13      349892  59.0   70.0  424.0  55.0  1.7  138.0  4.5  12.6  37.0  ...   
..         ...   ...    ...    ...   ...  ...    ...  ...   ...   ...  ...   
379     486397  20.0   70.0  123.0  44.0  1.0  135.0  3.8  14.6  44.0  ...   
381     462892  36.0   80.0   85.0  16.0  1.1  142.0  4.1  15.6  44.0  ...   
386     475200  64.0  100.0  163.0  54.0  7.2  140.0  4.6   7.9  26.0  ...   
387     168242  48.0   60.0  112.0  44.0  1.2  142.0  4.9  14.5  44.0  ...   
388     474407  21.0   90.0  107.0  40.0  1.7  125.0  3.5   8.3  23.0  ...   

     

## Part 3.4: Sort the dataset according to the Labels

In [6]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_data = cleaned_data.sort_values(by='Target').reset_index(drop=True)
# Print the dataset
print(sorted_data.head())

   unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1     137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2     484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3     343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4     532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   

         pc         pcc          ba  htn   dm  cad appet   pe  ane Target  
0  abnormal     present     present   no   no   no  good   no  yes    ckd  
1  abnormal     present  notpresent  yes  yes  yes  poor   no   no    ckd  
2  abnormal     present  notpresent  yes  yes   no  poor   no  yes    ckd  
3  abnormal  notpresent  notpresent  yes  yes  yes  poor  yes  yes    ckd  
4    normal  notpresent  notpresent   no   no   no  good   no   no    ckd  

[5 rows x 25 columns]


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [7]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encoded_data = pd.get_dummies(sorted_data, columns=categorical_columns, drop_first=True)

# Print the dataset
print(encoded_data.head())

   unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1     137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2     484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3     343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4     532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   

   pc_normal  pcc_present  ba_present  htn_yes  dm_yes  cad_yes  appet_poor  \
0      False         True        True    False   False    False       False   
1      False         True       False     True    True     True        True   
2      False         True       False     True    True    False        True   
3      False        False       False     True    True     True        True   
4       True        False       False    False   False    False       False   

   pe_yes  ane_yes  Target_notckd  
0   False     True        

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [8]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for col in numerical_columns:
    col_mean = encoded_data[col].mean()
    col_std = encoded_data[col].std()
    encoded_data = encoded_data[(encoded_data[col] >= col_mean - 3 * col_std) & (encoded_data[col] <= col_mean + 3 * col_std)]
# Print the dataset
print(encoded_data)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
3       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
6       546225  46.0   60.0  163.0   92.0  3.3  141.0  4.0   9.8  28.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     647552  61.0   70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
149     125587  37.0   60.0  109.0   47.0  1.1  141.0  4.9  15.0  48.0  ...   
150     614376  57.0   80.0  119.0   17.0  1.2  135.0  4.7  15.4  42.0  ...   
151     853426  41.0   70.0  125.0   38.0  0.6  140.0  5.0  16.8  41.0  ...   
152     572140  55.0   80.0  104.0   28.0  0.9  142.0  4.8  17.3  52.0  ...   

     pc_normal  pcc_present  ba_present  htn_yes  d

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [11]:
# Normalize the all Numerical Attributes in the dataset.
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
encoded_data[numerical_columns] = scaler.fit_transform(encoded_data[numerical_columns])
# Print the dataset
print(encoded_data.head())

   unique_id       age    bp       bgr        bu        sc       sod  \
0     474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
1     137148  0.905405  1.00  0.965665  0.522876  0.641975  0.666667   
3     343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
4     532520  0.729730  0.75  0.150215  0.281046  0.234568  0.533333   
6     546225  0.540541  0.00  0.399142  0.535948  0.358025  0.700000   

        pot      hemo       pcv  ...  pc_normal  pcc_present  ba_present  \
0  0.206897  0.059406  0.000000  ...      False         True        True   
1  0.000000  0.148515  0.225806  ...      False         True       False   
3  0.793103  0.000000  0.032258  ...      False        False       False   
4  0.793103  0.336634  0.322581  ...       True        False       False   
6  0.379310  0.207921  0.161290  ...       True        False       False   

   htn_yes  dm_yes  cad_yes  appet_poor  pe_yes  ane_yes  Target_notckd  
0    False   False    False       Fa

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: unique_id is not appropriate for modeling since it only serves as a primary key(an identifier). It does not actually provide any meaningful information for the model. There is no relevence in predicting the dataset.

In [12]:
#Remove that column
encoded_data.drop(columns=['unique_id'], inplace=True)
# Print the dataset
print(encoded_data.head())

        age    bp       bgr        bu        sc       sod       pot      hemo  \
0  0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897  0.059406   
1  0.905405  1.00  0.965665  0.522876  0.641975  0.666667  0.000000  0.148515   
3  0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103  0.000000   
4  0.729730  0.75  0.150215  0.281046  0.234568  0.533333  0.793103  0.336634   
6  0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.379310  0.207921   

        pcv      wbcc  ...  pc_normal  pcc_present  ba_present  htn_yes  \
0  0.000000  0.653226  ...      False         True        True    False   
1  0.225806  0.217742  ...      False         True       False     True   
3  0.032258  0.395161  ...      False        False       False     True   
4  0.322581  0.500000  ...       True        False       False    False   
6  0.161290  0.830645  ...       True        False       False     True   

   dm_yes  cad_yes  appet_poor  pe_yes  ane_yes  Target_notckd

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [18]:
# Export the dataframe to a new csv file
encoded_data.to_csv("cleaned_data.csv", index=False)
# Export the dataframe to a new json file
encoded_data.to_json("cleaned_data.json")

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

### Paste the markdown table here
|     |    age |    bp |       bgr |       bu |       sc |       sod |       pot |       hemo |       pcv |       wbcc |   ... |   pc_normal |   pcc_present |   ba_present |   htn_yes |   dm_yes |   cad_yes |   appet_poor |   pe_yes |   ane_yes |   Target_notckd |
|-----|--------|-------|-----------|----------|----------|-----------|-----------|------------|-----------|------------|-------|-------------|---------------|--------------|-----------|----------|-----------|---------------|----------|-----------|-----------------|
|  0  | 0.202703|  0.75 | 0.158798  | 0.196078 | 0.160494 | 0.166667  | 0.206897  | 0.059406   | 0.000000  | 0.653226   | ...   |      False  |         True   |        True  |     False |   False  |     False  |        False   |   False  |      True |           False |
|  1  | 0.905405|  1.00 | 0.965665  | 0.522876 | 0.641975 | 0.666667  | 0.000000  | 0.148515   | 0.225806  | 0.217742   | ...   |      False  |         True   |       False  |      True |     True  |      True  |         True   |   False  |     False |           False |
|  3  | 0.743243|  0.50 | 0.442060  | 0.901961 | 0.432099 | 0.500000  | 0.793103  | 0.000000   | 0.032258  | 0.395161   | ...   |      False  |        False   |       False  |      True |     True  |      True  |         True   |   True   |      True |           False |
|  4  | 0.729730|  0.75 | 0.150215  | 0.281046 | 0.234568 | 0.533333  | 0.793103  | 0.336634   | 0.322581  | 0.500000   | ...   |       True  |        False   |       False  |     False |   False  |     False  |        False   |   False  |     False |           False |
|  6  | 0.540541|  0.00 | 0.399142  | 0.535948 | 0.358025 | 0.700000  | 0.379310  | 0.207921   | 0.161290  | 0.830645   | ...   |       True  |        False   |       False  |      True |     True  |     False  |        False   |   False  |     False |           False |
|  8  | 0.675676|  0.75 | 0.253219  | 0.633987 | 0.777778 | 0.366667  | 0.655172  | 0.138614   | 0.193548  | 0.169355   | ...   |      False  |        False   |       False  |      True |   False  |     False  |        False   |   False  |     False |           False |
| 10  | 0.716216|  0.50 | 1.000000  | 0.163399 | 0.111111 | 0.066667  | 0.206897  | 0.267327   | 0.387097  | 0.532258   | ...   |       True  |        False   |       False  |     False |    True  |     False  |         True   |   False  |     False |           False |
| 13  | 0.729730|  0.00 | 0.935622  | 0.169935 | 0.160494 | 0.333333  | 0.034483  | 0.019802   | 0.064516  | 0.879032   | ...   |      False  |         True   |       False  |      True |   False  |     False  |        False   |   False  |      True |           False |
| 17  | 0.000000|  0.00 | 0.103004  | 0.372549 | 0.074074 | 0.500000  | 0.689655  | 0.217822   | 0.225806  | 1.000000   | ...   |      False  |        False   |        True  |     False |   False  |     False  |        False   |   False  |     False |           False |
| 19  | 0.878378|  0.00 | 0.206009  | 0.751634 | 0.604938 | 0.533333  | 0.689655  | 0.366337   | 0.387097  | 0.879032   | ...   |       True  |        False   |       False  |      True |    True  |     False  |         True   |    True  |     False |           False |
| 20  | 0.851351|  0.25 | 0.618026  | 0.562092 | 0.728395 | 0.000000  | 0.344828  | 0.168317   | 0.161290  | 0.580645   | ...   |      False  |         True   |        True  |      True |    True  |      True  |        False   |    True  |      True |           False |
| 21  | 0.878378|  0.25 | 0.639485  | 0.470588 | 0.395062 | 0.433333  | 0.517241  | 0.267327   | 0.322581  | 0.104839   | ...   |      False  |         True   |        True  |      True |    True  |      True  |        False   |   False  |     False |           False |
| 22  | 0.783784|  0.00 | 0.725322  | 0.313725 | 0.481481 | 0.566667  | 0.862069  | 0.178218   | 0.193548  | 0.258065   | ...   |      False  |        False   |        True  |      True |    True  |     False  |         True   |    True  |      True |           False |
| 24  | 0.662162|  0.50 | 0.618026  | 0.411765 | 0.432099 | 0.566667  | 0.689655  | 0.316832   | 0.354839  | 0.250000   | ...   |      False  |         True   |        True  |      True |    True  |     False  |        False   |    True  |     False |           False |
| 25  | 0.770270|  1.00 | 0.901288  | 0.163399 | 0.345679 | 0.766667  | 0.206897  | 0.524752   | 0.548387  | 0.443548   | ...   |       True  |        False   |        True  |      True |   False  |      True  |        False   |   False  |     False |           False |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

To mitigate hallucinations, break the data into smaller chunks and verify the outputs with the original dataset. You could also simplify tasks by asking for specific steps like generating code.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [15]:
# Code to rename all the columns in the dataset
import pandas as pd

# Sample DataFrame creation (modify this to load your actual dataset)
data = {
    'age': [0.202703, 0.905405, 0.743243],
    'bp': [0.75, 1.00, 0.50],
    'sg': ['1.020', '1.025', '1.010'],
    'al': ['0', '1', '0'],
    'su': ['0', '0', '0'],
    'rbc': [0, 1, 0],
    'pc': [0, 0, 1],
    'pcc': [1, 0, 0],
    'ba': [0, 1, 0],
    'bgr': [0.158798, 0.965665, 0.442060],
    'bu': [0.196078, 0.522876, 0.901961],
    'sc': [0.160494, 0.641975, 0.432099],
    'sod': [0.166667, 0.666667, 0.500000],
    'pot': [0.206897, 0.000000, 0.793103],
    'hemo': [0.059406, 0.148515, 0.000000],
    'pcv': [0.000000, 0.225806, 0.032258],
    'wbcc': [0.653226, 0.217742, 0.395161],
    'rbcc': [0.000000, 0.000000, 0.007068],  # Not listed previously
    'htn': [0, 1, 0],
    'dm': [0, 1, 1],
    'cad': [0, 1, 0],
    'appet': [0, 1, 0],
    'pe': [0, 0, 1],
    'ane': [1, 0, 1],
    'class': [0, 1, 1],
}

df = pd.DataFrame(data)

# Column renaming mapping based on your specifications
rename_mapping = {
    'age': 'age',
    'bp': 'blood_pressure',
    'sg': 'specific_gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red_blood_cells',
    'pc': 'pus_cell',
    'pcc': 'pus_cell_clumps',
    'ba': 'bacteria',
    'bgr': 'blood_glucose_random',
    'bu': 'blood_urea',
    'sc': 'serum_creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed_cell_volume',
    'wbcc': 'white_blood_cell_count',
    'rbcc': 'red_blood_cell_count',
    'htn': 'hypertension',
    'dm': 'diabetes_mellitus',
    'cad': 'coronary_artery_disease',
    'appet': 'appetite',
    'pe': 'pedal_edema',
    'ane': 'anemia',
    'class': 'ckd_or_not_ckd'
}

# Apply the renaming
df.rename(columns=rename_mapping, inplace=True)

# Display the renamed DataFrame
print(df.head())


        age  blood_pressure specific_gravity albumin sugar  red_blood_cells  \
0  0.202703            0.75            1.020       0     0                0   
1  0.905405            1.00            1.025       1     0                1   
2  0.743243            0.50            1.010       0     0                0   

   pus_cell  pus_cell_clumps  bacteria  blood_glucose_random  ...  \
0         0                1         0              0.158798  ...   
1         0                0         1              0.965665  ...   
2         1                0         0              0.442060  ...   

   packed_cell_volume  white_blood_cell_count  red_blood_cell_count  \
0            0.000000                0.653226              0.000000   
1            0.225806                0.217742              0.000000   
2            0.032258                0.395161              0.007068   

   hypertension  diabetes_mellitus  coronary_artery_disease  appetite  \
0             0                  0              

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [17]:
# Converted SQL to Pandas code
import pandas as pd

# Sample DataFrame creation (modify this to load your actual dataset)
data = {
    'Target': [0, 1, 0, 1, 1, 0, 1, 0, 0, 1],
}

df = pd.DataFrame(data)

# Group by 'Target' and count occurrences
result = df.groupby('Target').size().reset_index(name='count')

# Display the result
print(result)


   Target  count
0       0      5
1       1      5


*Describe what the above code does here*
The final output shows the number of occurrences of each unique value in the Target col. The result is a DataFrame with 2 cols.